File: python/1218.py
# Link: https://leetcode.com/problems/longest-arithmetic-subsequence-of-given-difference/description/
# Runtime: 524 ms
# Memory: 30 MB

class Solution:
    def longestSubsequence(self, arr: List[int], difference: int) -> int:
        length_of_subsequence_ended_by_value = defaultdict(lambda: 0)
        for val in arr:
            length_of_subsequence_ended_by_value[val] = length_of_subsequence_ended_by_value[val - difference] + 1

        return max(length_of_subsequence_ended_by_value.values())

File: c++/1218.cpp
// Link: https://leetcode.com/problems/longest-arithmetic-subsequence-of-given-difference/description/
// Runtime: 151 ms
// Memory: 56.52 MB

class Solution {
public:
  int longestSubsequence(const std::vector<int> &arr, int difference) {
    int maxSubsequence = 1;
    std::unordered_map<int, int> lengthOfSubsequenceEndedByValue;
    for (const auto &val : arr) {
      lengthOfSubsequenceEndedByValue[val] =
          lengthOfSubsequenceEndedByValue[val - difference] + 1;
      maxSubsequence =
          std::max(maxSubsequence, lengthOfSubsequenceEndedByValue[val]);
    }

    return maxSubsequence;
  }
};
